File: benchmarks/cross-language/elixir/thread_ring.exs
# Elixir Thread Ring Benchmark (Savina-style)

defmodule ThreadRing do
  @ring_size 100
  @num_hops (System.get_env("BENCHMARK_MESSAGES") |> case do
    nil -> 100_000
    val -> String.to_integer(val)
  end)

  def start do
    IO.puts("=== Elixir Thread Ring Benchmark ===")
    IO.puts("Ring size: #{@ring_size}, Hops: #{@num_hops}\n")

    parent = self()

    # Create ring of processes
    first = spawn(fn -> ring_node(nil, parent, 0) end)
    last = create_ring(first, @ring_size - 1, parent)

    # Close the ring: last -> first (not first -> last!)
    send(last, {:set_next, first})

    start_time = :erlang.monotonic_time(:nanosecond)

    # Send initial token
    send(first, {:token, @num_hops})

    # Wait for completion
    receive do
      {:done, total_received} ->
        end_time = :erlang.monotonic_time(:nanosecond)
        elapsed_ns = end_time - start_time
        elapsed_sec = elapsed_ns / 1_000_000_000.0

        # Each node receives ~(num_hops + 1) / ring_size messages
        expected_per_node = div(@num_hops + 1, @ring_size) + 1
        if total_received < expected_per_node - 10 or total_received > expected_per_node + 10 do
          IO.puts("VALIDATION WARNING: expected ~#{expected_per_node} per node, got #{total_received}")
        end

        throughput = (@num_hops + 1) / elapsed_sec / 1_000_000
        ns_per_msg = elapsed_ns / (@num_hops + 1)

        IO.puts("ns/msg:         #{:io_lib.format("~.2f", [ns_per_msg])}")
        IO.puts("Throughput:     #{:io_lib.format("~.2f", [throughput])} M msg/sec")
    after
      60_000 ->
        IO.puts("Timeout!")
        System.halt(1)
    end
  end

  defp create_ring(prev, 0, _parent), do: prev
  defp create_ring(prev, n, parent) do
    next = spawn(fn -> ring_node(prev, parent, 0) end)
    send(prev, {:set_next, next})
    create_ring(next, n - 1, parent)
  end

  defp ring_node(next, parent, received) do
    receive do
      {:set_next, new_next} ->
        ring_node(new_next, parent, received)
      {:token, 0} ->
        send(parent, {:done, received + 1})
      {:token, hops} ->
        send(next, {:token, hops - 1})
        ring_node(next, parent, received + 1)
    end
  end
end

ThreadRing.start()

File: benchmarks/cross-language/erlang/thread_ring.erl
-module(thread_ring).
-export([start/0]).

-define(RING_SIZE, 100).

get_num_hops() ->
    case os:getenv("BENCHMARK_MESSAGES") of
        false -> 100000;
        Val -> list_to_integer(Val)
    end.

start() ->
    NumHops = get_num_hops(),
    io:format("=== Erlang Thread Ring Benchmark ===~n"),
    io:format("Ring size: ~p, Hops: ~p~n~n", [?RING_SIZE, NumHops]),

    % Create ring of processes
    Parent = self(),
    First = spawn(fun() -> ring_node(undefined, Parent, 0) end),
    Last = create_ring(First, ?RING_SIZE - 1, Parent),

    % Close the ring: Last -> First (not First -> Last!)
    Last ! {set_next, First},

    % Start timer (monotonic_time is correct for benchmarks)
    Start = erlang:monotonic_time(nanosecond),

    % Send initial token
    First ! {token, NumHops},

    % Wait for completion
    receive
        {done, TotalReceived} ->
            End = erlang:monotonic_time(nanosecond),
            ElapsedNs = End - Start,
            ElapsedSec = ElapsedNs / 1000000000.0,

            % Each node receives ~(NumHops + 1) / RING_SIZE messages
            ExpectedPerNode = (NumHops + 1) div ?RING_SIZE + 1,
            if
                TotalReceived < ExpectedPerNode - 10; TotalReceived > ExpectedPerNode + 10 ->
                    io:format("VALIDATION WARNING: expected ~p per node, got ~p~n", [ExpectedPerNode, TotalReceived]);
                true -> ok
            end,

            NsPerMsg = ElapsedNs / (NumHops + 1),
            MsgPerSec = (NumHops + 1) / ElapsedSec,
            io:format("ns/msg:         ~.2f~n", [NsPerMsg]),
            io:format("Throughput:     ~.2f M msg/sec~n", [MsgPerSec / 1000000])
    after 60000 ->
        io:format("Timeout!~n"),
        halt(1)
    end,

    halt(0).

create_ring(Prev, 0, _Parent) ->
    Prev;
create_ring(Prev, N, Parent) ->
    Next = spawn(fun() -> ring_node(Prev, Parent, 0) end),
    Prev ! {set_next, Next},
    create_ring(Next, N - 1, Parent).

ring_node(Next, Parent, Received) ->
    receive
        {set_next, NewNext} ->
            ring_node(NewNext, Parent, Received);
        {token, 0} ->
            % Token exhausted, report to parent
            Parent ! {done, Received + 1};
        {token, Hops} ->
            Next ! {token, Hops - 1},
            ring_node(Next, Parent, Received + 1)
    end.
